File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
import java.util.regex.*;

class Solution {
    public String solution(String m, String[] musicinfos) {
        m = m.replace("C#","c").replace("D#","d").replace("F#","f").replace("G#","g").replace("A#","a");
        Pattern pattern = Pattern.compile(m);

        
        int playTime = 0;
        String answer = "(None)";
        
        for (String music : musicinfos) {
            String[] s = music.split(",");
            int minute = changeMinutes(s[0], s[1]);
            String name = s[2];
            String sound = s[3].replace("C#","c").replace("D#","d").replace("F#","f").replace("G#","g").replace("A#","a");
            
            Matcher matcher = pattern.matcher(sound.repeat(minute/sound.length()) + sound.substring(0,minute % sound.length()));
            if (matcher.find() && minute > playTime) {
                playTime = minute;
                answer = name;
            }
        }
        return answer;
    }
    
    public int changeMinutes(String start, String end) {
        String[] s = start.split(":");
        String[] e = end.split(":");
        
        return (Integer.parseInt(e[0]) - Integer.parseInt(s[0])) * 60 + Integer.parseInt(e[1]) - Integer.parseInt(s[1]);
    }
}

File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.py
import re

import heapq
def solution(m, musicinfos):
    m = m.replace("C#", "c").replace("D#", "d").replace("F#", "f").replace("G#", "g").replace("A#", "a")
    
    pattern = re.compile(m)
    answer = []
    
    for i in range(len(musicinfos)):
        s = musicinfos[i].split(",")
        minute = changeMinute(s[0], s[1])
        name = s[2]
        music = s[3].replace("C#", "c").replace("D#", "d").replace("F#", "f").replace("G#", "g").replace("A#", "a")
        
        if (pattern.search(music * (minute // len(music)) + music[:minute % len(music)])):
            answer.append([minute, name])
    
    return "(None)" if not answer else max(answer, key = lambda x: x[0])[1]
        
    
        
        
        
def changeMinute(start, end):

    s_h, s_m = map(int, start.split(":"))
    e_h, e_m = map(int, end.split(":"))
    
    return (e_h - s_h) * 60 + (e_m - s_m)
    
    
    
        
        